lib/curl_sha512_256.c

```diff
@@ -0,0 +1,588 @@
+/***************************************************************************
+ *                                  _   _ ____  _
+ *  Project                     ___| | | |  _ \| |
+ *                             / __| | | | |_) | |
+ *                            | (__| |_| |  _ <| |___
+ *                             \___|\___/|_| \_\_____|
+ *
+ * Copyright (C) Evgeny Grin (Karlson2k), <k2k@narod.ru>.
+ *
+ * This software is licensed as described in the file COPYING, which
+ * you should have received as part of this distribution. The terms
+ * are also available at https://curl.se/docs/copyright.html.
+ *
+ * You may opt to use, copy, modify, merge, publish, distribute and/or sell
+ * copies of the Software, and permit persons to whom the Software is
+ * furnished to do so, under the terms of the COPYING file.
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
+ * KIND, either express or implied.
+ *
+ * SPDX-License-Identifier: curl
+ *
+ ***************************************************************************/
+
+#include "curl_setup.h"
+
+#if !defined(CURL_DISABLE_DIGEST_AUTH) && !defined(CURL_DISABLE_SHA512_256)
+
+#include "curl_sha512_256.h"
+#include "warnless.h"
+
+/* ** This implementation of SHA-512/256 hash calculation was originally ** *
+ * ** written by Evgeny Grin (Karlson2k) for GNU libmicrohttpd.          ** *
+ * ** The author ported the code to libcurl. The ported code is provided ** *
+ * ** under curl license.                                                ** *
+ * ** This is a minimal version with minimal optimisations. Performance  ** *
+ * ** can be significantly improved. Big-endian store and load macros    ** *
+ * ** are obvious targets for optimisation.                              ** */
+
+#if !defined(MHDX_UINT64) && defined(SIZEOF_LONG)
+#if SIZEOF_LONG == 8
+#define MHDX_UINT64 unsigned long
+#define MHDX_UINT64_C(val) val##UL
+#endif /* SIZEOF_LONG == 8 */
+#endif /* !MHDX_UINT64 && SIZEOF_LONG */
+
+#if !defined(MHDX_UINT64) && defined(HAVE_LONGLONG)
+#define MHDX_UINT64 unsigned long long
+#define MHDX_UINT64_C(val) val##ULL
+#endif /* !MHDX_UINT64 && SIZEOF_LONG */
+
+#if !defined(MHDX_UINT64)
+#define MHDX_UINT64 unsigned __int64
+#define MHDX_UINT64_C(val) ((MHDX_UINT64)val)
+#endif /* !MHDX_UINT64 && SIZEOF_LONG */
+
+#ifndef MHDX_UINT64_TRIM
+/* Trim possibly larger value to 64 bits.
+   Assuming that no type is larger than 64 bits. */
+#define MHDX_UINT64_TRIM(x) ((MHDX_UINT64)(x))
+#endif /* !MHDX_UINT64_TRIM */
+
+/* Bits manipulation macros and functions.
+   Can be moved to other headers to reuse. */
+
+#define MHDX_GET_64BIT_BE(ptr)                                \
+  MHDX_UINT64_TRIM(                                           \
```

Comment:
When you OR 64 bits together, it can't be larger than 64 bits so there's no need to trim this down to 64 bits. Even if the type could potentially hold something larger.

Reply:
Right. I'll correct it.